### PR TITLE
Adding @shopify/graphql-persisted

### PR DESCRIPTION
**Pull request by lemonmade**
Now that we almost have the necessary tooling done for persisted queries (https://github.com/Shopify/sewing-kit/pull/1195), it's time to do the application code side. This PR introduces a new library, `graphql-persisted`. This library provides an Apollo Link and Koa middleware that work together to make persisted queries work, and also updates `sewing-kit-koa` to support the new GraphQL manifest from sewing kit.

Basically, the process is this:

1. On the client, the Apollo Link sees that the GraphQL document has an `id` field. It sets this on the `extensions.persisted.id` field of the request, and tells the HTTP link not to send the query body.
2. On the server, the Koa middleware parses the body to grab that ID. If it's found, it first tries to ask `sewing-kit-koa` for the full source that maps to that ID. If that's not found, it instead asks an optional cache (allowing teams to store this information in redis or somewhere else that makes sense).
3. If a document is found using these techniques, the middleware sets `request.body.query` to be the full query body and exits. Subsequent middleware will use that (it's the same field on the request body that they would have received for a non-persisted query).
4. If no such document is found, the behaviour depends on the option they pass for `cacheMissBehavior`. The default is that the server will ask the client to retry, but to always send the full query for that ID, instead of just the ID. The consumer of this package can choose instead to ask the client to retry + store the new mapping in the cache, or error straight up.

### CODE REVIEW

**Comment by lemonmade** [comment]
No tests yet, will work on them once people have gotten a chance to feel out the approach itself.

**Comment by GoodForOneFare** [comment]
If you're wondering _y tho?_, the Apollo docs say:
> 1. Restricting the queries that your server accepts by whitelisting a specific set of queries
> 2. Saving bandwidth between the client and server by only sending a query ID or hash instead of an entire GraphQL query

1 is compelling.  2 seems eh, but I don't know enough about our query sizes to make a useful judgement.

**Comment by lemonmade** [comment]
It's a good point Gord! Not sure whether Web will restrict queries, at least at first, but 2 I think will have enough of a benefit on its own. Upload bandwidth is always slowest, and on a home load of the admin, I saw 7.5kb of query bodies in the outgoing requests. Granted, this will be reduced significantly by the new loaders I wrote, it will still be substantially larger than the <1kb that those same queries would have been if they only sent their IDs.

**Comment by GoodForOneFare** [comment]
Proving that I'm definitely not a good judge of query sizes 😆 Thanks, for the extra context 👍

File: packages/graphql-persisted/src/apollo.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        },
+      });
+
+      let subscription: ReturnType<
+        ReturnType<typeof forward>['subscribe']
+      > | null = null;
+
+      subscription = forward(operation).subscribe({
+        next: response => {
+          const errors = response ? response.errors || [] : [];
```
Comment (GoodForOneFare) [inline packages/graphql-persisted/src/apollo.ts]:
```suggestion
          const errors = response && response.errors || [];
```

But I don't mind either way ¯\\_(ツ)_/¯

File: packages/graphql-persisted/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+const client = new ApolloClient({
+  /* other options, like the cache */
+  link: ApolloLink.from([createPersistedLink(), createHttpLink()]),
+});
+```
+
+This function accepts an optional `options` object. The following option is available:
+
+- `idFromOperation?(operation: Operation): string | undefined | null`: calculates the unique ID to use for the persisted query, which will eventually be passed to the server’s `cache#get` method to retrieve the full query body. If omitted, this option will default to pulling the `id` field off of the `operation.query` `DocumentNode`, which works well in combination with documents compiled using [`graphql-mini-transforms`](https://github.com/Shopify/graphql-tools-web/tree/master/packages/graphql-mini-transforms) (used by default in sewing kit).
```
[the remark this answers is not in the document]
Comment (lemonmade) [reply packages/graphql-persisted/README.md]: It's not a typo, I just don't like using it in plain English sentences, personally. I'll change it for (slightly more) consistency.

File: packages/graphql-persisted/src/tests/koa-middleware.test.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    });
+
+    await persistedGraphQLMiddleware(ctx, spy);
+
+    expect(spy).toHaveBeenCalled();
+    expect(cache.get).not.toHaveBeenCalled();
+    expect(cache.set).not.toHaveBeenCalled();
+  });
+
+  it('does nothing if there is a persisted query body and explicit GraphQL query', async () => {
```
Comment (cartogram) [inline packages/graphql-persisted/src/tests/koa-middleware.test.ts]: Not a big fan of this test title because you are asserting on the same things as the previous test, which is not nothing.

File: packages/graphql-persisted/src/tests/koa-middleware.test.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    const spy = jest.fn();
+    const persistedGraphQLMiddleware = createPersistedGraphQLMiddleware();
+
+    await persistedGraphQLMiddleware(ctx, spy);
+
+    expect(spy).toHaveBeenCalled();
+    expect(ctx.request).toHaveProperty('body.query', query);
+  });
+
+  it('saves query returned by sewing-kit-koa to the cache when provided', async () => {
```
Comment (cartogram) [inline packages/graphql-persisted/src/tests/koa-middleware.test.ts]: `saves the query`

File: packages/graphql-persisted/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  }),
+);
+
+// additional code, including whatever middleware will actually
+// handle the GraphQL query
+```
+
+When the middleware attempts to resolve an ID to a query document, it attempts to do the following:
+
+- If `ctx.state.assets.graphQLSource` exists, call it and wait on the result. If a result is found and a cache was passed, return the result _and_ save it to the cache using `cache.set`. This step makes it automatic to get persisted queries going for apps using sewing kit and [`@shopify/sewing-kit-koa`](https://github.com/Shopify/quilt/tree/master/packages/sewing-kit-koa).
```
[the remark this answers is not in the document]
Comment (cartogram) [reply packages/graphql-persisted/README.md]: I added this comment a while ago and some stuff was shifted around, it should be referring to line `77`.

File: packages/graphql-persisted/src/apollo.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
+import {ApolloLink, Operation, NextLink, Observable} from 'apollo-link';
+import {CacheMissBehavior} from './shared';
+
+interface Options {
+  idFromOperation?(operation: Operation): string | undefined | null;
+}
+
+export function createPersistedLink(options: Options = {}) {
```
[the remark this answers is not in the document]
Comment (lemonmade) [reply packages/graphql-persisted/src/apollo.ts]: Apollo does this for its other links also so mostly just for consistency.
